lib/ansible/utils/vars.py

```diff
@@ -114,13 +113,6 @@ def merge_hash(x, y, recursive=True, list_merge='replace'):
     # we don't want to modify x, so we create a copy of it
     x = x.copy()
 
-    # to speed things up: use dict.update if possible
-    # (this `if` can be remove without impact on the function
-    #  except performance)
-    if not recursive and list_merge == 'replace':
-        x.update(y)
-        return x
-
```

Comment:
It might be possible to restore this optimization by keeping track of whether merge tags have been used.